### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by kashike**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1209) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/680)

---

This is a work-in-progress pull request, and the description will be updated once more progress has been made.

### CODE REVIEW

**Comment by liach** [comment]
What kind of settings?

File: src/main/java/org/spongepowered/api/setting/Setting.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.setting;
+
+public class Setting {
```
Comment (liach) [inline src/main/java/org/spongepowered/api/setting/Setting.java]: Possibly change this to an interface or a catalog type

**Comment by Zidane** [comment]
@kashike 

While this seems obvious, I'm having issues seeing the point of this when we have configurate. Can you go into detail?

File: src/main/java/org/spongepowered/api/setting/value/EnumSettingValue.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public String serialize() {
+        return this.value == null ? "" : this.value.name();
+    }
+
+    @Override
+    public Optional<T> cycleNext() {
+        T[] constants = this.enumClass.getEnumConstants();
+        if (constants.length == 0) {
```
Comment (Electroid) [inline src/main/java/org/spongepowered/api/setting/value/EnumSettingValue.java]: It does not make sense for any setting value to have no possible options, this should be a precondition in the object. That also means you can remove the `Optional` return from `CycleableSettingValue#cycleNext`. If you can cycle something, there should always be a valid setting option.

Comment (kashike) [inline src/main/java/org/spongepowered/api/setting/value/EnumSettingValue.java]: Good point, will do that now.

**Comment by gabizou** [comment]
I'm somewhat confused as to the relevance of this PR. That aside, there's a lot of implementation detail that could reside in, well, the implementation of Sponge API. Additionally, making Setting not a `CatalogType` is somewhat silly if the sake of having generic type information is to be desired.

**Comment by kashike** [comment]
@gabizou What do you think can go into the implementation? `SimpleSettingManager` and `SimpleSettingRegistry` can be used by plugins to construct their own registries and managers, `*SettingType` can be extended. `SimpleSetting` can be moved to the implementation if really desired.

**Comment by kashike** [comment]
I'm going to close this for now - I'll revisit this soon.

**Pull request closed without merging**